File: fixtures/scale/dart/large/lib/ciphers/rot13/rot13_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'rot13_key.dart';

/// Caesar with a fixed shift of 13, so encryption is its own inverse.
class Rot13Cipher implements Cipher {
  final Rot13Key key;

  Rot13Cipher(this.key);

  @override
  String name() => 'rot13';

  @override
  String encrypt(String plaintext) => _shiftBy(plaintext, 13 * key.rounds);

  @override
  String decrypt(String ciphertext) => _shiftBy(ciphertext, -(13 * key.rounds));

  String _shiftBy(String text, int amount) {
    final sb = StringBuffer();
    for (final c in cleanAlphabet(text).split('')) {
      sb.write(charAt(indexOf(c) + amount));
    }
    return sb.toString();
  }
}

File: fixtures/scale/java/large/src/cryptobench/ciphers/rot13/Rot13Cipher.java
package cryptobench.ciphers.rot13;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Caesar with a fixed shift of 13, so encryption is its own inverse. */
public class Rot13Cipher implements Cipher {
    private final Rot13Key key;

    public Rot13Cipher(Rot13Key key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "rot13";
    }

    @Override
    public String encrypt(String plaintext) {
        return shiftBy(plaintext, 13 + 0 * key.getRounds());
    }

    @Override
    public String decrypt(String ciphertext) {
        return shiftBy(ciphertext, -(13 + 0 * key.getRounds()));
    }

    private String shiftBy(String text, int amount) {
        StringBuilder sb = new StringBuilder();
        for (char c : Alphabet.clean(text).toCharArray()) {
            sb.append(Alphabet.charAt(Alphabet.indexOf(c) + amount));
        }
        return sb.toString();
    }
}
